core/test/base/math.cpp

```diff
@@ -0,0 +1,99 @@
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/base/math.hpp>
+
+
+#include <cmath>
+#include <complex>
+#include <limits>
+
+
+#include <gtest/gtest.h>
+
+
+namespace {
+
+
+template <typename T>
+void test_real_isfinite()
+{
+    using limits = std::numeric_limits<T>;
+    constexpr auto inf = limits::infinity();
+
+    ASSERT_TRUE(gko::isfinite(T{0}));
+    ASSERT_TRUE(gko::isfinite(-T{0}));
+    ASSERT_TRUE(gko::isfinite(T{1}));
+    ASSERT_FALSE(gko::isfinite(inf));
+    ASSERT_FALSE(gko::isfinite(-inf));
+    ASSERT_FALSE(gko::isfinite(limits::quiet_NaN()));
+    ASSERT_FALSE(gko::isfinite(limits::signaling_NaN()));
+}
+
+
+template <typename T>
+void test_complex_isfinite()
+{
+    using limits = std::numeric_limits<T>;
+    using c_type = std::complex<T>;
+    constexpr auto inf = limits::infinity();
+    constexpr auto quiet_nan = limits::quiet_NaN();
+    constexpr auto signaling_nan = limits::signaling_NaN();
+
+    ASSERT_TRUE(gko::isfinite(c_type{T{0}, T{0}}));
+    ASSERT_TRUE(gko::isfinite(c_type{-T{0}, -T{0}}));
+    ASSERT_TRUE(gko::isfinite(c_type{T{1}, T{0}}));
+    ASSERT_TRUE(gko::isfinite(c_type{T{0}, T{1}}));
+    ASSERT_FALSE(gko::isfinite(c_type{inf, T{0}}));
+    ASSERT_FALSE(gko::isfinite(c_type{-inf, T{0}}));
+    ASSERT_FALSE(gko::isfinite(c_type{quiet_nan, T{0}}));
+    ASSERT_FALSE(gko::isfinite(c_type{signaling_nan, T{0}}));
+    ASSERT_FALSE(gko::isfinite(c_type{T{0}, inf}));
+    ASSERT_FALSE(gko::isfinite(c_type{T{0}, -inf}));
+    ASSERT_FALSE(gko::isfinite(c_type{T{0}, quiet_nan}));
+    ASSERT_FALSE(gko::isfinite(c_type{T{0}, signaling_nan}));
+}
+
+
+TEST(IsFinite, Float) { test_real_isfinite<float>(); }
```

Comment:
The test is only meant for definitions of `isfinite` inside `math.hpp`, which currently only supports `float`, `double` and both in complex. If we add a new format to it, we definitively need to test that here as well. But in order to test it, I would need to know what custom precision, since they can differ a lot (and blindly testing some does not really help).

Reply:
But we do have stuff like [`half`](https://github.com/ginkgo-project/ginkgo/blob/metis-reorder/include/ginkgo/core/base/math.hpp#L141) and and [`truncated`](https://github.com/ginkgo-project/ginkgo/blob/metis-reorder/include/ginkgo/core/base/math.hpp#L256), right ?

Reply:
Yes, we do, but currently, we never use `isfinite` in combination with it (as far as I know, it is only used in factorization / ILU).
But I think when someone needs it, it can be added by them, since these are edge cases and might need special optimizations.